src/orange/OrangeTrackView.hh

```diff
@@ -491,19 +551,80 @@ CELER_FUNCTION void OrangeTrackView::set_dir(Real3 const& newdir)
         {
             // The boundary crossing direction has changed! Reverse our plans
             // to change the logical state and move to a new volume.
-            states_.boundary[thread_]
-                = flip_boundary(states_.boundary[thread_]);
+            lsa.boundary() = flip_boundary(lsa.boundary());
         }
     }
 
     // Complete direction setting
-    states_.dir[thread_] = newdir;
+    lsa.dir() = newdir;
 
     this->clear_next_step();
 }
 
 //---------------------------------------------------------------------------//
 // PRIVATE MEMBER FUNCTIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Iterate over all levels to find the next step.
+ */
+CELER_FUNCTION void OrangeTrackView::find_next_step_impl(real_type max_step)
+{
+    UniverseId min_uid;
+    celeritas::detail::OnSurface min_surface_local;
+    auto min_step = max_step;
+
+    // Find the nearest intersection from level 0 to current level inclusive,
+    // prefering the higher level (i.e., lowest uid)
+    for (auto level : range(states_.level[thread_] + 1))
+    {
+        auto lsa = this->make_lsa(LevelId{level});
+        auto tracker = this->make_tracker(lsa.universe());
+        auto isect = tracker.intersect(this->make_local_state(LevelId{level}),
+                                       max_step);
+
+        if (isect.distance < min_step)
+        {
```

Comment:
I forgot that since SCALE ORANGE doesn't have the `max step` argument, we can get a big optimization here: instead of limiting the intersection search length to `max_step`, limit it to `min_step`. In the 'complex volume' case this will result in a huge speedup because fewer surfaces will have to be stored in global memory and sorted.

With this in place you can also eliminate the `min_step` variable (or rename `max_step` to `min_step` locally: `max_step` is the maximum search distance, `min_step` is the smallest distance to a boundary, and they should both be the same).

Reply:
Good catch on the optimization. I think having the variable named "min_step" makes more sense. With this in mind I could either do:

```
find_next_step_impl(real_type min_step)
```

as you suggest, or alternatively 

```
find_next_step_impl(real_type max_step)
{
...
min_step = max_step
...
}
```
I kind of like this alternative even thought it introduces an additional variable, as it "black boxes" the algorithm within `find_next_step_impl`, i.e., when calling `find_next_step_impl`, the caller should not be required to understand how the underlying algorithm works. I think the method is more easier understood if the caller is asked to supply a "max_step", since this is what it actually is. The fact that this max_step is minimized over the course of the algorithm is in am implementation detail in my opinion. What do you think?

Reply:
Maybe `search_step`? Also since the declaration variable names don't have to match the definition, you could use `max_step` in the former and `min_step` in the latter 😅 

I suspect the compiler will be able to optimize whatever choice we take, so I'm kind of in agreement with you... if you don't like `search_step` we can do two variables.

Reply:
I think having different names in the declaration/definition seems strange and is probably more confusing that just using min_dist in both places. `search_step` is not bad, but I prefer `max_step` because its analogous to `SimpleUnitTracker.insersect(state, max_dist)`